engine/src/main/java/org/hibernate/search/indexes/impl/PeriodicRegreshingReaderProvider.java

```diff
@@ -0,0 +1,301 @@
+		log.tracef( "Opening IndexReader for directoryProvider %s", indexName );
+		Directory directory = directoryProvider.getDirectory();
+		PerDirectoryLatestReader directoryLatestReader = currentReaders.get( directory );
+		// might eg happen for FSSlaveDirectoryProvider or for mutable SearchFactory
+		if ( directoryLatestReader == null ) {
+			directoryLatestReader = createReader( directory );
+		}
+		return directoryLatestReader.getLatestReader();
+	}
+
+	@Override
+	public void closeIndexReader(IndexReader reader) {
+		if ( reader == null ) {
+			return;
+		}
+		log.tracef( "Closing IndexReader: %s", reader );
+		ReaderUsagePair container = allReaders.get( reader );
+		container.close(); //virtual
+	}
+
+	//overridable method for testability:
+	protected DirectoryReader readerFactory(final Directory directory) throws IOException {
+		return DirectoryReader.open( directory );
+	}
+
+	@Override
+	public void initialize(DirectoryBasedIndexManager indexManager, Properties props) {
+		this.directoryProvider = indexManager.getDirectoryProvider();
+		this.indexName = indexManager.getIndexName();
+		// Initialize at least one, don't forget directoryProvider might return a different Directory later
+		createReader( directoryProvider.getDirectory() );
+		this.scheduledExecutorService = Executors.newScheduledThreadPool( "Periodic IndexReader refreshing task for index " + indexName );
+		this.delay = ConfigurationParseHelper.getIntValue(
+				props,
+				Environment.ASYNC_READER_REFRESH_PERIOD_MS,
+				DEFAULT_ACCEPTABLE_MAX_AGE_MS );
+		this.scheduledExecutorService.scheduleWithFixedDelay(
+				new IndexRefreshTask(),
+				delay,
+				delay,
+				TimeUnit.MILLISECONDS
+		);
+	}
+
+	@Override
+	public void stop() {
+		if ( scheduledExecutorService != null ) {
+			try {
+				scheduledExecutorService.shutdown();
+				scheduledExecutorService.awaitTermination( 1, TimeUnit.SECONDS );
+				scheduledExecutorService = null;
+			}
+			catch (InterruptedException e) {
+				log.timedOutWaitingShutdownOfReaderProvider( indexName );
+			}
+		}
+		for ( IndexReader reader : allReaders.keySet() ) {
+			ReaderUsagePair usage = allReaders.get( reader );
+			usage.close();
+		}
+
+		if ( allReaders.size() != 0 ) {
+			log.readersNotProperlyClosedInReaderProvider();
+		}
+	}
+
+
+	/**
+	 * Thread safe creation of <code>PerDirectoryLatestReader</code>.
+	 *
+	 * @param directory The Lucene directory for which to create the reader.
+	 * @return either the cached instance for the specified <code>Directory</code> or a newly created one.
+	 * @see <a href="http://opensource.atlassian.com/projects/hibernate/browse/HSEARCH-250">HSEARCH-250</a>
+	 */
+	private synchronized PerDirectoryLatestReader createReader(Directory directory) {
+		PerDirectoryLatestReader reader = currentReaders.get( directory );
+		if ( reader != null ) {
+			return reader;
+		}
+
+		try {
+			reader = new PerDirectoryLatestReader( directory );
+			currentReaders.put( directory, reader );
```

Comment:
Right, in this case if there is a concurrent request it's preferrable to have a thread wait on the synchronization's lock rather than doing multiple IO operations to attempt to open multiple readers (and then have to close and discard one of them).

Think of it in terms of spent resources: if you lock, the thread will be parked and CPU time assigned to do something useful on a different thread. The cost is if course that your current thread will have to wait N milliseconds, were N is the time for the other thread to complete the indexreader refresh.

In the alternative strategy, using a putIfAbsent, the thread will not hit the lock but it will have to perform the same IO operations, which will take it N milliseconds to do (as it's the same operation as in the previous case).

So if you compare the two scenarios, while the lock might looks like "contention", it actually frees the CPU but does less IO, and takes the take time for the client to actually get an IndexReader.

So that's why I choose for the synchronized block.